utt/utt-client-api/include/utt-client-api/User.hpp

```diff
@@ -0,0 +1,140 @@
+// UTT Client API
+//
+// Copyright (c) 2020-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <vector>
+#include <string>
+#include <variant>
+#include <optional>
+
+// [TODO-UTT] All types are tentative
+
+namespace utt::client {
+
+enum class Error {};
+struct TransferTx {};
+struct MintTx {};
+struct BurnTx {};
+
+using Tx = std::variant<TransferTx, BurnTx, MintTx>;
+using BurnResult = std::variant<TransferTx, BurnTx, Error>;
+using TransferResult = std::variant<TransferTx, Error>;
+
+class User {
+ public:
+  /**
+   * @brief Create the user registration commitment "rcm1".
+   *
+   * @return rcm1 as a byte array
+   */
+  std::vector<uint8_t> getRegistrationInput() const;
+
+  /**
+   * @brief Use a registration with the following credentials:
+   *
+   * @param pk The user's public key associated with the registration
+   * @param rs The signature on user's registration commitment
+   * @param s2 The system-side part of the user's PRF key
+   *
+   * @return true if the use of the registration is accepted
+   */
+  bool useRegistration(const std::string& pk, const std::vector<uint8_t>& rs, const std::vector<uint8_t>& s2);
+
+  /**
+   * @brief Use a previously created budget coin for anonymous transactions
+   *
+   * @param budgetCoin The budget coin
+   *
+   * @return true if the use of the budget coin is accepted
+   */
+  bool useBudgetCoin(const std::vector<uint8_t>& budgetCoin);
+
+  // [TODO-UTT] Decide if we need a separate interface for setting the budget coin and sig.
+  /**
+   * @brief Use a signature for a previously created budget coin
+   *
+   * @param budgetCoin The budget coin
+   *
+   * @return true if the use of the budget coin signature is accepted
+   */
+  bool useBudgetCoinSig(const std::vector<uint8_t>& sig);
+
+  /**
+   * @brief Get the total value of unspent UTT tokens
+   */
+  uint64_t getBalance() const;
+
+  /**
+   * @brief Get the value of the currently avialable budget coin
+   */
+  uint64_t getBudget() const;
```

Comment:
`getPrivacyBudget()` ?

Reply:
A more general comment is that "budget" and "balance" are a bit difficult to distinguish, unless maybe we specify it is a "privacy budget".